src/cesride.udl

```diff
@@ -0,0 +1,322 @@
+[Error]
+enum Error {
+    "General",
+    "Matter",
+    "EmptyMaterial",
+    "Decode",
+    "UnexpectedCode",
+    "UnexpectedCountCode",
+    "UnexpectedOpCode",
+    "InvalidVarSize",
+    "InvalidVarRawSize",
+    "InvalidVarIndex",
+    "InvalidCodeSize",
+    "InvalidBase64Character",
+    "InvalidBase64Index",
+    "Shortage",
+    "EmptyQb64",
+    "UnknownSizage",
+    "UnknownHardage",
+    "UnknownBardage",
+    "UnsupportedSize",
+    "TooSmall",
+    "Prepad",
+    "NonZeroedPrepad",
+    "NonZeroedLeadByte",
+    "NonZeroedLeadBytes",
+    "NonZeroedPadBits",
+    "Parsing",
+    "ParseQb64",
+    "ParseQb2",
+    "Conversion",
+    "Value",
+    "Validation",
+    "Derivation"
+};
+
+enum CesrideMatterCodex {
```

Comment:
This is no longer an enum, it is simple constants.

Reply:
Yes, I know. We did that to simplify things, since making the code go through the ffi destroys the abilty to call 'code' on it if it is an enum. With the way you've defined it, the wrapping methods need the enum and can't be passed a string. I guess the concern is - will a consumer need to instantiate a class with a string? The other problem is that in the future the code tables will not be fixed to one set of static codes, we'll have loadable codes and this strategy means we'd need to create a new library for each consumer, rather than letting them define their code tables externally, load them into `cesride/parside`, and simply play with text strings. They'd need to create a new UDL, write new wrappers etc.

@m00sey what do you think? I don't want us to get boxed in by needing to support that kind of workflow

Reply:
But you are right, because of not using enums we have dead code where we match `_` and we may miss one in a match.

Reply:
I guess one other option is that we could refactor so that the loadable tables are Rust crates, and they have a UDL as well. They'd need to conform to a certain signature, and I think that's the tricky bit where I'm not sure it works.

What I think may work with more confidence is changing these methods to accept strings and defining the code tables in the consuming bindings - for example you probably have convenience wrappers in kotlin that create something like classes to encapsulate functionality defined by the functions exposed below. What do you think about putting the codes there, turning the inputs here into strings and removing the enums? I'd like to know what @m00sey thinks of this idea, too.

Reply:
I'm thinking through the signature thing I spoke of - you'd need to have validation groupings at minimum for all concrete classes, in the code table. I don't see why this isn't possible though? Is there anything else that would need to be defined?

Reply:
Probably everything in `tables`.

Reply:
I think we could make a trait from it though. I'm going to write a few lines of code and paste it here.

Reply:
So, here is what I came up with:
https://github.com/jasoncolburne/cesride/pull/1/files

I think it can probably work but I still don't know that this is the correct choice.

Reply:
It's obviously incomplete, but it demonstrates how we could make a conditionally compiled use statement that pulled different Enums conforming to that trait into the lib for different use cases

Reply:
for example building with a particular table would just be a matter of turning on a feature and having the dependency available

Reply:
> I did not find a way to define a constant through FFI.

I feel I have reached a point where I cannot approve these PRs without investigating other FFI options. I need to better understand the trade offs from multiple FFI implementations (or at least one additional) before making concessions.

@jasoncolburne has done a lot of work to "try make uniffi work" so I am going to take some time to look at RustJNI https://google.github.io/comprehensive-rust/android/interoperability/java.html

Until then I feel we should pause review of the parside/cesride/uniffi work.

Once we have a better understanding we can pick a direction as a community.

Reply:
I will clarify this is my _opinion_ and I want this to be a discussion within the community.